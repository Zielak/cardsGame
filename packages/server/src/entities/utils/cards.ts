import { logs } from "@cardsgame/utils"

import type { ClassicCard } from "../classicCard.js"

/**
 * All basic ranks in an array (order matters)
 */
// prettier-ignore
export const ALL_CARD_RANKS = [
  "2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K", "A"
]

/**
 * All basic suits in an array (order matters)
 */
export const ALL_CARD_SUITS = ["D", "S", "H", "C"]

/**
 * Will generate an array of card options.
 * Use this array to create actual cards yourself
 * @example
 * ```ts
 * standardDeckFactory().map(options => {
 *   new ClassicCard({state, ...options})
 * })
 * ```
 *
 * @param ranks array of desired ranks
 * @param suits array of desired suits
 */
export const standardDeckFactory = (
  // prettier-ignore
  ranks: string[] = ALL_CARD_RANKS,
  suits: string[] = ALL_CARD_SUITS
): Pick<ClassicCard, "rank" | "suit">[] => {
  const cards: Pick<ClassicCard, "rank" | "suit">[] = []

  for (const suit of suits) {
    for (const rank of ranks) {
      cards.push({ suit, rank })
    }
  }

  logs.log(`created data for a deck of ${cards.length} cards`)

  return cards
}
