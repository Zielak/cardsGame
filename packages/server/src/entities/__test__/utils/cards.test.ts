import { standardDeckFactory } from "../../utils/cards.js"

describe("standardDeckFactory", () => {
  it("gives data for 52 cards", () => {
    const cards = standardDeckFactory()

    expect(cards.length).toBe(52)
    expect(cards.some((el) => el === undefined)).toBeFalsy()
    expect(cards.every((el) => typeof el.suit === "string")).toBeTruthy()
    expect(cards.every((el) => typeof el.rank === "string")).toBeTruthy()
  })

  it("respects arrays", () => {
    const cards = standardDeckFactory(["2", "3"], ["H", "C"])

    expect(cards.length).toBe(4)
    expect(cards).toMatchSnapshot()
  })

  it("respects sets", () => {
    const cards = standardDeckFactory(new Set(["2", "3"]), new Set(["H", "C"]))

    expect(cards.length).toBe(4)
    expect(cards).toMatchSnapshot()
  })
})
