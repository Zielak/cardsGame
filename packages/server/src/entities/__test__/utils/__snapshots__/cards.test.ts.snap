// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`standardDeckFactory respects arrays 1`] = `
[
  {
    "rank": "2",
    "suit": "H",
  },
  {
    "rank": "3",
    "suit": "H",
  },
  {
    "rank": "2",
    "suit": "C",
  },
  {
    "rank": "3",
    "suit": "C",
  },
]
`;
